src/cdomains/mval.ml

```diff
@@ -0,0 +1,74 @@
+include Mval_intf
+
+open GoblintCil
+
+module M = Messages
+
+
+module MakePrintable (Offs: Offset.Printable): Printable with type idx = Offs.idx =
+struct
+  type idx = Offs.idx
+  include Printable.StdLeaf
+  (* Use Basetype.Variables to print with RichVarinfo. *)
+  type t = Basetype.Variables.t * Offs.t [@@deriving eq, ord, hash]
+
+  let name () = Format.sprintf "lval (%s)" (Offs.name ())
+
+  let show ((v, o): t): string = CilType.Varinfo.show v ^ Offs.show o
+  include Printable.SimpleShow (
+    struct
+      type nonrec t = t
+      let show = show
+    end
+    )
+
+  let add_offset (v, o) o' = (v, Offs.add_offset o o')
+
+  let type_of (v,o) = try Offs.type_of ~base:v.vtype o with Offset.Type_of_error (t,_) -> t
+
+  let prefix (v1,ofs1) (v2,ofs2) =
+    if CilType.Varinfo.equal v1 v2 then
+      Offs.prefix ofs1 ofs2
+    else
+      None
+
+  let to_cil ((v, o): t): lval = (Var v, Offs.to_cil o)
+  let to_cil_exp lv = Lval (to_cil lv)
+
+  let is_definite (_, o) = Offs.is_definite o
```

Comment:
Maybe call `is_definite_offset`? This will return true also for heap-allocated things etc which might be confusing.

Reply:
That's true, although in `AddressDomain` `is_definite` isn't just about offsets again, so it would make for inconsistent naming. There we also consider `NullPtr` to be definite. So it seems that the semantics of this check has never been very well defined.